REGRESSION (283478@main): [ Sonoma arm64 wk2 Release ] 4x http/tests/webgpu/webgpu/* tests are flaky failing

mwyrzykowski opened the pull request:
#### 93be927de1ed6471954cb09a1f8d5669b2f8ee41
<pre>
REGRESSION (283478@main): [ Sonoma arm64 wk2 Release ] 4x http/tests/webgpu/webgpu/* tests are flaky failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=279769">https://bugs.webkit.org/show_bug.cgi?id=279769</a>
<a href="https://rdar.apple.com/136087702">rdar://136087702</a>

Reviewed by Cameron McCormack.

283478@main introduced a retain cycle between Device and XRSubImage.

Break the retain cycle similarly to what we do for Queue: use a WeakPtr
in the non-Device class.

* Source/WebGPU/WebGPU/XRSubImage.h:
* Source/WebGPU/WebGPU/XRSubImage.mm:
(WebGPU::XRSubImage::XRSubImage):
(WebGPU::XRSubImage::update):
Break retain cycle

* LayoutTests/platform/mac-wk2/TestExpectations
Undo skipped expectations.

Canonical link: <a href="https://commits.webkit.org/283746@main">https://commits.webkit.org/283746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f66c425a2d794457d7a9c4a939d907266c1c6ea7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67231 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46610 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19863 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71271 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18364 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18158 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53880 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12336 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70298 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42842 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58167 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34414 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39514 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15560 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16720 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61479 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15901 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72975 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11190 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15243 "Found 1 new test failure: media/video-replaces-poster.html (failure)") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61362 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11223 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58225 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61449 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14906 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9182 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2774 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43492 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43233 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebGPU/WebGPU/XRSubImage.mm:
```diff
@@ -66,6 +66,11 @@
+    RefPtr device = m_device.get();
+    auto* devicePtr = device.get();
+    if (!devicePtr)
+        return;
```
Is there a need to store the raw pointer in a local variable? I would instead just do:

```
if (!device)
    return;
...
auto newTexture = Texture::create(..., *device);
...
```

heycam commented on Source/WebGPU/WebGPU/XRSubImage.mm:
```diff
@@ -36,12 +36,12 @@
-    : m_device(device)
+    : m_device(&device)
 {
 }
```
I think `ThreadSafeWeakPtr` has a constructor that works with a const reference to an object that inherits from `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr`, so you might not need the `&`s.

mwyrzykowski committed:
REGRESSION (283478@main): [ Sonoma arm64 wk2 Release ] 4x http/tests/webgpu/webgpu/* tests are flaky failing
https://bugs.webkit.org/show_bug.cgi?id=279769
rdar://136087702

Reviewed by Cameron McCormack.

283478@main introduced a retain cycle between Device and XRSubImage.

Break the retain cycle similarly to what we do for Queue: use a WeakPtr
in the non-Device class.

* Source/WebGPU/WebGPU/XRSubImage.h:
* Source/WebGPU/WebGPU/XRSubImage.mm:
(WebGPU::XRSubImage::XRSubImage):
(WebGPU::XRSubImage::update):
Break retain cycle

* LayoutTests/platform/mac-wk2/TestExpectations
Undo skipped expectations.

Canonical link: https://commits.webkit.org/283746@main
